include/circt/Support/PrettyPrinter.h

```diff
@@ -72,34 +73,66 @@ class Token {
   struct EndInfo : public TokenInfo {
     // Nothing
   };
+  // This can be used to associate a callback with the print event on the
+  // tokens stream. Note that the lambda used to create the function object will
+  // be out of scope when it is evoked. So extra care is required to ensure the
+  // values captured by the function object are valid out of scope.
+  struct CallbackInfo : public TokenInfo {
+    using CallbackTy = std::function<void()>;
+    CallbackTy callback;
+  };
 
 private:
-  union {
-    TokenInfo info;
-    StringInfo stringInfo;
-    BreakInfo breakInfo;
-    BeginInfo beginInfo;
-    EndInfo endInfo;
-  } data;
+  std::variant<StringInfo, BreakInfo, BeginInfo, EndInfo, CallbackInfo> data;
```

Comment:
Token classes should be entirely reworked if you want to move to std::variant, which I don't think is a good idea (`std::variant<>`) but definitely shouldn't be done like this (LLVM RTTI, redundant tracking of tag, so on).

Having tokens larger and require destruction / no-longer-POD does not seem like a win, and doesn't change what's going on with heap allocations for these `std::function`s.